### PR TITLE
5.2 Merge

### PR DESCRIPTION
**Pull request by ncik-roberts**
Merge with a version of the work-in-progress OCaml 5.2 branch.

Branch I merged with: https://github.com/ocaml-flambda/flambda-backend/tree/merge-5.2 as of commit 33aedfc93c38ccad7a4d89974405c05123a18932

## Overall plan

* This PR brings merlin-jst in line with merge-5.2
* A subsequent PR will also merlin-jst with upstream merlin's main branch. This will allow us to import changes and improvements made upstream, and will make our lives easier for future merges with upstream/main.

We will merge the stacked PR into this branch before merging it into main. In particular, the tests fail in this PR (though the tree builds).

## Suggestions for review
* Skip the first commit. It runs the standard import script to update the upstream copy of the code.
* Skip the second commit. It commits the merge conflicts created by pulling the upstream copy of the code into the tree.
* Review the third commit. It does the tricky job of actually merging in the changes from flambda-backend. I generally tended to take the flambda-backend side of the diff, but this isn't universally true.

## Note for reviewer

This isn't the final version of this PR, but it can be reviewed in its current state (and I will merge additional changes to the merge-5.2 branch into this PR before we merge it, which will create a little more review).

### CODE REVIEW

**Review by ccasin — APPROVED** [review]
Looks good, just a question and a nit.

File: src/ocaml/typing/typecore.ml
Comment (ncik-roberts) [reply src/ocaml/typing/typecore.ml:1062]: Good catch, thanks. I think calls to `unify_pat_types` are guarded with try-with, though one would be forgiven for not spotting that.

File: src/ocaml/preprocess/menhirLib.ml
Comment (ncik-roberts) [reply src/ocaml/preprocess/menhirLib.ml:3795]: I think I should try to keep this at 20210419, per Liam's comment here: https://github.com/janestreet/merlin-jst/pull/86

**Comment by ncik-roberts** [comment]
Thanks for the review. Obviously, it conflicts terribly with the latest merlin roll. This is what I'm going to do:
1. Wait until the flambda-backend 5.2 merge branch has progressed to a better state.
2. Re-import the upstream copy of flambda-backend in this PR and fix conflicts.
3. Merge with origin/main and fix conflicts.

This is a lot of work, so I'm going to hold off until we feel that this PR is close to mergeable.

File: upstream/ocaml_flambda/utils/config.common.ml
Comment (liam923) [inline upstream/ocaml_flambda/utils/config.common.ml:1]: I think something got messed up here

**Review by liam923 — CHANGES_REQUESTED** [review]

File: upstream/ocaml_flambda/utils/config.common.ml
Comment (liam923) [reply upstream/ocaml_flambda/utils/config.common.ml:1]: `config.common.ml` got replaced with `config.common.ml.in` in flambda. I deleted this file, and added `upstream/ocaml_flambda/configure`, which is the file that now contains the magic numbers

**Review by liam923 — APPROVED** [review]

**Comment by liam923** [comment]
After merging, we realized it was a mistake to do a squash-merge instead of a normal merge. We re-wrote history for it to be a normal merge instead - see PR #98

**Pull request merged**
